### PR TITLE
Bug fix: Disable filament shrink compensation for multicolor/material prints which was causing incorrect colour segmentation.

### LINKED ISSUE #4091: Filament shrinkage is mixing an objects' coloring
### OrcaSlicer Version

1.9.0

### OS version

Windows 11

### Additional system information

_No response_

### Printer

Bambu Lab X1C

### How to reproduce

1. Open an object that is suitable for multi color
2. Color with two colors (black and red for example)
3. Slice the object and verify all looks well
4. Enter to one of the selected colors filament settings
5. Set the shrinkage value, in my case 99.97% for PLA
6. Slice the object again

### Actual results

The slicer mixes the coloring.

image

### Expected results

Filament shrinkage should not affect the object multi-coloring

image

### Project file & Debug log uploads

[two-faced-sturdy-infinity-cube-20.zip](https://github.com/SoftFever/OrcaSlicer/files/14279627/two-faced-sturdy-infinity-cube-20.zip)

### Checklist of files to include

- [ ] Log file
- [X] Project file

### LINKED ISSUE #4551: Sliced multi-color model inncorrect color on top layer
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### OrcaSlicer Version

2.0.0

### Operating System (OS)

Linux

### OS Version

Mint 21.3

### Additional system information

image

### Printer

Bambu X1C

### How to reproduce

Open attached model of lid riser, model is blue with silver bambu logo.  Slice model, expect blue model with silver logog.

### Actual results

Logo sliced with blue filament

### Expected results

Logo should print with silver filament

### Project file & Debug log uploads

[Bambu Top Lid Glass Riser_Orca_Slice_Bug.zip](https://github.com/SoftFever/OrcaSlicer/files/14637899/Bambu.Top.Lid.Glass.Riser_Orca_Slice_Bug.zip)

### Checklist of files to include

- [X] Log file
- [X] Project file

### Anything else?

Left: Preview of blue model with silver logo
Middle: Sliced with Bambu studio, blue model with silver logo (expected result)
Right: Sliced with Orca, blue model with blue logo (wrong, logo should have top layer of silver).

Screenshot from 2024-03-18 07-34-15

### LINKED ISSUE #5152: Multi color painting only applies to top/bottom shell layers
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### OrcaSlicer Version

2.0.0

### Operating System (OS)

Windows

### OS Version

Windows 11

### Additional system information

CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz, 2592 Mhz, 6 Core(s), 12 Logical Processor(s)
Memory: 16GB Physical, 27.8 Virtual
GPU: Nvidia RTX 2060

### Printer

Voron 2.4

### How to reproduce

1. Import Test Object
2. Paint surfaces for different filament colors (Circle, Sphere, and/or Fill)
3. Slice

### Actual results

Object is sliced with filament changes only applied to top and bottom shell layers, not vertical shells.
mal sliced
test sliced

### Expected results

Objects should be sliced with surface colors matching painting performed.
Mal prep
test prepare

### Project file & Debug log uploads

[MMU Malenia bug.zip](https://github.com/SoftFever/OrcaSlicer/files/15118274/MMU.Malenia.bug.zip)

### Checklist of files to include

- [ ] Log file
- [X] Project file

### Anything else?

Edit: Changing shrinkage percentage for all filaments to 100% resolves this issue

### LINKED ISSUE #5664: Slicing issue colour loss
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### OrcaSlicer Version

2.1beta

### Operating System (OS)

macOS

### OS Version

Monterey 12.6.8

### Additional system information

_No response_

### Printer

Bambu a1

### How to reproduce

IMG_6517
Skull has black eyes for multi colour but this is removed when slicing 

### Actual results

IMG_6516

### Expected results

Eyes should stay black

### Project file & Debug log uploads

[orca skull colour issue.3mf.zip](https://github.com/user-attachments/files/15784979/orca.skull.colour.issue.3mf.zip)

### Checklist of files to include

- [ ] Log file
- [X] Project file

### Anything else?

_No response_

### LINKED ISSUE #5756: Multicolor not slicing correctly
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### OrcaSlicer Version

2.0 - 2.1 rc

### Operating System (OS)

Linux

### OS Version

manjaro

### Additional system information

_No response_

### Printer

voron 2.4

### How to reproduce

open stl, paint something , slice

### Actual results

image

### Expected results

image

### Project file & Debug log uploads

[Voron_Design_Cube_v8.zip](https://github.com/user-attachments/files/15892838/Voron_Design_Cube_v8.zip)

### Checklist of files to include

- [ ] Log file
- [ ] Project file

### Anything else?

Quick test with latest appimage of bamboo studio
image

### PR DESCRIPTION
**Pull request by igiannakas**
# Description

Fixes the below issues reported when using filament shrink compensation together with a multi-color/material print. 

The fix is detecting the use of multiple extruders and disables the compensation for multi material/color prints as it is not compatible. This is similar to the XY hole compensation that is correctly also disabled when printing multi-material prints. 

This has been reported several times on both discord and here so maybe this correction helps.

Fixes #5756
Fixes #5152
Fixes #4551
Fixes #5664
Fixes #4091

image

## Tests

Tested with multi colour print with shrink compensation - slices successfully
Tested with single colour print with shrink compensation - slices successfully and applies XY correction.

### CODE REVIEW

**Comment by BioCookieYT** [comment]
Thanks for the workaround!
I'm wondering if there is another way to achieve filament shrinkage correction since the current way has some issues with some slicer features. An idea would be to simply use the already build in object scaling function as a shrinkage correction:
` -> check if shrinkage is activated -> scale object using object scaling -> run slicing process`

image



From my testing scaling doesn't affect MMU painting:
image
image


I'm not really in the C++ game so not sure if something like this would be possible. Just some idea since scaling is basically filament shrinkage correction from my understanding.

**Comment by igiannakas** [comment]
The problem is that scaling may also vary by filament. One could be 0.9 and the other 0.95 or something. Then what do you use? 🤔

**Comment by BioCookieYT** [comment]
Good point 🤔
There are 3 ways I can think of:

1. Cancel slicing with the error "Filament shrinkage must be equal" 
2. Calculate the average shrinkage from all used filaments 
3. Use the shrinkage of the "main filament" (in the cube example the black one -> the filament which was used first) 

It kinda depends on the situation, for parts with just some colored text I would use option 3 or 1, for parts with like 4 colors in one layer option 2. Maybe an option for this in the Multicolor printer tab would be useful? (probably a lot of work to implement😐)

**Comment by igiannakas** [comment]
It’s the same issue as XY hole compensation that is skipped when doing a multi color print. In my view, material shrinkage should be baked into the model by design, so I’ve never used this. And if needed we can just adjust the size of the model as a complete scale factor. 

However, many ppl have this set up and then try to slice a multi material print and then it fails and they don’t know why. So I’m proposing this approach for now, to protect the user from using an incompatible feature. 

I’ll investigate if there is an alternative way but not very hopeful as this feature was lifted from SS, which also has this issue, and neither Bambu nor prusa have this implemented, so we cannot cross check what the “upstream” branches have done.

**Comment by SoftFever** [comment]
I think it would be better to have a warning notification instead of silently disable shrinkage

**Comment by igiannakas** [comment]
> I think it would be better to have a warning notification instead of silently disable shrinkage

I tried that in a similar manner that XY compensation does but the method to raise the slicing message is protected and can’t call it from the parameter pointer :(. Any idea on what pop up to use as an alternative?

**Comment by igiannakas** [comment]
Mmm let me try something different

**Comment by igiannakas** [comment]
Hey @SoftFever  I've implemented the warning message in two places - one triggered at slice time and one at the validate function. Pick your favourite and I'll remove the other :) 

The slice time message takes you to the multicoloured object, so that's nice; however it needs me to mark a function as public which I'm not too keen. 

Pick one and I'll cleanup the code removing the other one.

File: src/libslic3r/PrintObjectSlice.cpp [hunk 1]
```diff
@@ -447,22 +447,47 @@ static std::vector<std::vector<ExPolygons>> slices_to_regions(
...
+            } else if (scale != 1) {
                 scale = 1 / scale;
                 for (ExPolygons& polys : region_polys)
                     for (ExPolygon& poly : polys)
                         poly.scale(scale);
             }
+            
         }
     }
+    if(display_shrinkage_comp_error){
```
Comment (igiannakas) [inline src/libslic3r/PrintObjectSlice.cpp:487]: @SoftFever this is the slice time messsage here

File: src/libslic3r/Print.cpp [hunk 2]
```diff
@@ -1556,6 +1556,22 @@ StringObjectException Print::validate(StringObjectException *warning, Polygons*
...
+    if (warning && extruders.size() > 1){
+        bool filament_shrinkage_found = false;
+        for(int extruder : extruders){
+            if(m_config.filament_shrink.values[extruder] != 100){
+                filament_shrinkage_found = true;
+                break;
+            }
+        }
+        if (filament_shrinkage_found){
+            warning->string = L("Filament shrinkage compensation is incompatible with multi color and multi material prints. It will be ignored when printing this model.");
```
Comment (igiannakas) [inline src/libslic3r/Print.cpp:1570]: And the validate message here :)

**Comment by ExR90** [comment]
I'm having this issue, but my filaments are set to 100% for shrink already. Easy to reproduce, starting in version 2.1 (any, even Beta) create a primitive cube. Paint one vertical side with any color, then slice. That color won't be in the finished product. Reverting to 2.0 and it behaves as expected.

I don't have the ability to compile my own build to test your fix for my issue but I thought I would chime in mentioning I'm having this problem with the shrinkage already at 100.

**Comment by igiannakas** [comment]
> I'm having this issue, but my filaments are set to 100% for shrink already. Easy to reproduce, starting in version 2.1 (any, even Beta) create a primitive cube. Paint one vertical side with any color, then slice. That color won't be in the finished product. Reverting to 2.0 and it behaves as expected.
> 
> I don't have the ability to compile my own build to test your fix for my issue but I thought I would chime in mentioning I'm having this problem with the shrinkage already at 100.

Please save the sliced cube project file and upload it here in a zip file. 

I’ve done the exact same test as mentioned above and it doesn’t behave the same as yours.

**Comment by igiannakas** [comment]
You can download the ready made build here: https://github.com/SoftFever/OrcaSlicer/actions/runs/9615570849

bottom of the page.

**Comment by ExR90** [comment]
> > I'm having this issue, but my filaments are set to 100% for shrink already. Easy to reproduce, starting in version 2.1 (any, even Beta) create a primitive cube. Paint one vertical side with any color, then slice. That color won't be in the finished product. Reverting to 2.0 and it behaves as expected.
> > I don't have the ability to compile my own build to test your fix for my issue but I thought I would chime in mentioning I'm having this problem with the shrinkage already at 100.
> 
> Please save the sliced cube project file and upload it here in a zip file.
> 
> I’ve done the exact same test as mentioned above and it doesn’t behave the same as yours.

@igiannakas 

Video of the behavior in 2.0 and 2.1 from my POV. I am on MacOS 14.5 (23F79) on M3 CPU
https://share.zight.com/eDuKnZX6

Project file of the test cube and Gcode created of same cube is attached. 
[cube with 3 colors.zip](https://github.com/user-attachments/files/15936082/cube.with.3.colors.zip)

**Comment by ExR90** [comment]
> You can download the ready made build here: https://github.com/SoftFever/OrcaSlicer/actions/runs/9615570849
> 
> bottom of the page.

I tried this but cannot execute the binary, Finder tries to launch it and I get an error that the file is corrupted. It unzipped fine into the DMG, which also mounted fine and was able to extract the binary fine - it just states it is damaged when I try to run it.

Zight 2024-06-22 at 3 01 28 AM

**Comment by igiannakas** [comment]
> > You can download the ready made build here: https://github.com/SoftFever/OrcaSlicer/actions/runs/9615570849
> > bottom of the page.
> 
> I tried this but cannot execute the binary, Finder tries to launch it and I get an error that the file is corrupted. It unzipped fine into the DMG, which also mounted fine and was able to extract the binary fine - it just states it is damaged when I try to run it.
> 
> Zight 2024-06-22 at 3 01 28 AM

You need to run this command here to sign the file:
`xattr -dr com.apple.quarantine /Applications/OrcaSlicer.app`

**Comment by igiannakas** [comment]
Your project file attached above is slicing OK for me:
image

image


Honestly I'm lost as cannot reproduce it :( 

@SoftFever any thoughts?

**Comment by prnthp** [comment]
> You can download the ready made build here: https://github.com/SoftFever/OrcaSlicer/actions/runs/9615570849
> 
> bottom of the page.

Unfortunately this PR didn't seem to fix the issue.

Screenshot 2024-06-22 at 2 58 21 PM
Screenshot 2024-06-22 at 2 58 18 PM

Project file: https://github.com/prnthp/Cone/blob/main/Cone.3mf

**Comment by ExR90** [comment]
> Your project file attached above is slicing OK for me: image
> 
> image
> 
> Honestly I'm lost as cannot reproduce it :(
> 
> @SoftFever any thoughts?

I thought maybe there was something bitched up in the preferences, so I signed into my Mac with a new user, which gave me a brand new profile like on a fresh new install of Orca - it took me through the wizard and everything. I did not configure anything other than saying I had an X1C, then did the test shown above which appears to suggest the preferences are not storing something to cause the problem.

**Comment by ExR90** [comment]
@igiannakas 

Curious, I am on a Mac, and so is the other person that just chimed in on this thread. You on PC or Mac? Maybe there is something to that with respect to OS build?

**Comment by igiannakas** [comment]
I’m on an M1 Pro MacBook…

**Comment by ExR90** [comment]
> I’m on an M1 Pro MacBook…

No idea then. If I roll back to 2.0 it works fine. Something strange is afoot at the local Circle-K....

**Comment by igiannakas** [comment]
Just sliced the above model successfully with an M1 MacBook Pro. We need someone to debug this with an Intel Mac? Looks like this is happening with that only?

image

**Comment by igiannakas** [comment]
@SoftFever need help :) :) :) 

I've been trying to investigate this - but I'm hitting a brick wall.

**There are two issues here.** 

**Issue 1:** Shrink compensation, that is addressed with this PR successfully (pending comments above etc)

**Issue 2:** There seems to be a secondary issue that is present in the built files from the **GitHub actions:**

The below is taken from the build of **this PR**. Also it's the same with the **main release build**.
image

However when I build the **main branch locally with already built dependencies** all is good:
image

**Then I deleted my GitHub local folder including all build files etc and recloned the repo and rebuilt from scratch**. It still works properly even after a clean dependency rebuild etc.

I've also tried running a build using a slightly older version of the GitHub actions on my cloned main branch here:
https://github.com/igiannakas/OrcaSlicer/actions/workflows/build_all.yml

Firstly I downloaded my dev branch binaries from here:
https://github.com/igiannakas/OrcaSlicer/actions/runs/9632430284

These were built with dependencies cached around 2 weeks ago.
https://github.com/igiannakas/OrcaSlicer/actions/caches

**It doesnt work again...** So there is something up with the way the actions are building the binary for MacOS as it works with a local build but doesnt work with a GitHub actions build...

image

**Comment by Maximizer18** [comment]
> @SoftFever need help :) :) :)
> 
> I've been trying to investigate this - but I'm hitting a brick wall.
> 
> **There are two issues here.**
> 
> **Issue 1:** Shrink compensation, that is addressed with this PR successfully (pending comments above etc)
> 
> **Issue 2:** There seems to be a secondary issue that is present in the built files from the **GitHub actions:**
> 
> The below is taken from the build of **this PR**. Also it's the same with the **main release build**. image
> 
> However when I build the **main branch locally with already built dependencies** all is good: image
> 
> **Then I deleted my GitHub local folder including all build files etc and recloned the repo and rebuilt from scratch**. It still works properly even after a clean dependency rebuild etc.
> 
> I've also tried running a build using a slightly older version of the GitHub actions on my cloned main branch here: https://github.com/igiannakas/OrcaSlicer/actions/workflows/build_all.yml
> 
> Firstly I downloaded my dev branch binaries from here: https://github.com/igiannakas/OrcaSlicer/actions/runs/9632430284
> 
> These were built with dependencies cached around 2 weeks ago. https://github.com/igiannakas/OrcaSlicer/actions/caches
> 
> **It doesnt work again...** So there is something up with the way the actions are building the binary for MacOS as it works with a local build but doesnt work with a GitHub actions build...
> 
> image

Hi, so I'm encountering the same issue on Mac OS 12.7.5
Orca slicer version 2.0.0 works perfectly fine. I don't know the slightest bit about code, but can't you see what changes between the 2 versions to see what could be causing this?

**Comment by Maximizer18** [comment]
> Just sliced the above model successfully with an M1 MacBook Pro. We need someone to debug this with an Intel Mac? Looks like this is happening with that only?
> 
> image

I have an intel Mac. Tell me what I need to do to test whatever.

**Comment by ExR90** [comment]
> > Just sliced the above model successfully with an M1 MacBook Pro. We need someone to debug this with an Intel Mac? Looks like this is happening with that only?
> > 
> > image
> 
> I have an intel Mac. Tell me what I need to do to test whatever.

It isn't Intel or Mac silicon only. I've got M3 with the problem, others don't. As others have said it impacts any 2.1 version. All 2.0 versions and before slice fine. I've tried fresh install on fresh user profile and no change in behavior, as I had thought maybe something in the preferences was causing this but no.

**Comment by igiannakas** [comment]
Latest Mac build should solve the non shrinkage related MMU issue. Download from the nightly and use this command to "sign" it so it opens.

`xattr -dr com.apple.quarantine /Applications/OrcaSlicer.app`

**Comment by SoftFever** [comment]
I've updated the CI/CD build script, the problem should be fixed.

**Comment by Maximizer18** [comment]
Nightly build works! Thank you guys! Would the next update of the regular build include this fix?

**Comment by ExR90** [comment]
@igiannakas @SoftFever 

Just tried the nightly and it DOES solve the "no color on vertical surfaces" issue. Hooray!

**Comment by SoftFever** [comment]
Thank you @igiannakas 

I have two feedback:
1. I think a warning in Print::validate is sufficient.
2. It's probably better to check whether all used filaments have same shrinkage value? e.g. if all used filaments has 98% shrinkage, then they should just work fine.  There is a way to get what filaments are used in this print, I cant' remember whether the info is available before the validation check at the moment though.

**Comment by igiannakas** [comment]
I tried it by checking if the extruders have the same shrink comp value but the algo fails even if the shrink value is the same:

image
image

image

I've pushed the latest update removing the slice time message :)

**Comment by igiannakas** [comment]
My hunch is that the subsequent polygon segmentation ops dont like the polygons having moved. 

Also I've noticed that this:
const size_t extruder_id = pr->extruder(FlowRole::frPerimeter) - 1;
Returns all extruders when even one has shrink compensation enabled...

**Comment by SoftFever** [comment]
> My hunch is that the subsequent polygon segmentation ops dont like the polygons having moved.
> 
> Also if memory is not failing me, I've noticed that this: `const size_t extruder_id = pr->extruder(FlowRole::frPerimeter) - 1;` Returns all extruders when even one has shrink compensation enabled... Which is very very odd as I would have guessed the below correctly identifies the associated extruder but for some reason it appears not to.
> 
> ```
> unsigned int PrintRegion::extruder(FlowRole role) const
> {
>     size_t extruder = 0;
>     if (role == frPerimeter || role == frExternalPerimeter)
>         extruder = m_config.wall_filament;
>     else if (role == frInfill)
>         extruder = m_config.sparse_infill_filament;
>     else if (role == frSolidInfill || role == frTopSolidInfill)
>         extruder = m_config.solid_infill_filament;
>     else
>         throw Slic3r::InvalidArgument("Unknown role");
>     return extruder;
> }
> ```

I see.
Thank you for digging deep into the issue.
I'm thinking it might be better to port PrusaSlicer's new implementation when any of us have time: https://github.com/prusa3d/PrusaSlicer/commit/4b8fe98443b2d556c13653d3277b11e9b02c40a8

I feel that is a better implementation.
What do you think?

**Comment by igiannakas** [comment]
Oooh that looks nice! Not sure

**Comment by igiannakas** [comment]
Seems pretty straightforward to port - I’ll have a go at it next week/ the week after once I get a bit of free time from work :) 

closing this PR as there is a better way forward.

**Pull request closed without merging**